src/askui_runner/modules/core/infrastructure/files/askui.py

```diff
@@ -41,13 +57,158 @@ def __init__(self, base_url: str, headers: dict[str, str]):
         self.base_url = base_url.rstrip("/")
         self.headers = headers
 
+    # region Public Functions
+    def download(self, local_dir_path: str, remote_path: str = "") -> None:
+        """Download files from S3.
+
+        Args:
+            local_dir_path (str): The local directory to download the files to.
+            remote_path (str, optional): The remote path to a directory to download the files from or a single file to download. Defaults to "". If you pass a a prefix of a directory or file, the `remote_path` prefix is going to be stripped from the file paths when creating the local paths to the files.
+        """
+        if self.disabled:
+            return
+        prefix = remote_path.lstrip("/")
+        list_objects_response = self._list_object_recursive(prefix)
+        for content in list_objects_response.data:
+            if any(
+                re.match(pattern, content.path) is not None
+                for pattern in self.HIDDEN_FILES_PATTERNS
+            ):
+                continue
+
+            if prefix == content.path:  # is a file
+                relative_remote_path = content.name
+            else:  # is a prefix, e.g., folder
+                relative_remote_path = content.path[len(prefix) :].lstrip("/")
+            local_file_path = os.path.join(
+                local_dir_path, *relative_remote_path.split("/")
+            )
+            self._download_file(content.url, local_file_path, content.last_modified)
+
+    def upload(self, local_path: str, remote_dir_path: str = "") -> None:
+        if self.disabled:
+            return
+        r_dir_path = remote_dir_path.rstrip("/")
+        if os.path.isdir(local_path):
+            self._upload_dir(local_path, r_dir_path)
+        else:
+            self._upload_file(
+                local_path, f"{r_dir_path}/{os.path.basename(local_path)}"
+            )
+
+    def sync(
+        self,
+        local_dir_path: str,
+        remote_dir_path: str,
+        source_of_truth: Literal["local", "remote"],
+        dry: bool = False,
+        delete: bool = True,
+    ) -> None:
+        if source_of_truth not in {"local", "remote"}:
+            raise ValueError("source_of_truth must be 'local' or 'remote'")
+
+        # List remote files
+        remote_files: Dict[str, FileDto] = {}
+        list_objects_response = self._list_object_recursive(remote_dir_path)
+        for content in list_objects_response.data:
+            if any(
+                re.match(pattern, content.path)
+                for pattern in self.HIDDEN_FILES_PATTERNS
+            ):
+                continue
+            remote_files[content.path[len(remote_dir_path) :].lstrip("/")] = content
+
+        # List local files
+        local_files = self._list_local_files(local_dir_path)
+
+        # Create lookup table
+        all_paths = set(local_files.keys()) | set(remote_files.keys())
+
+        # Action mapper
+        action_mapper: Dict[str, Callable[..., None]] = {
```

Comment:
Do we need this action mapper? Can't we just add the `dry` as an argument defaulting to `False` to all the functions called and add the logging within them?

I think this could simplify the `sync` function